### PR TITLE
[vLLM] Speculative decode with NGram

### LINKED ISSUE #5706: [vLLM] Add support for NGram speculative decode

### PR DESCRIPTION
**Pull request by mmanzoorTT**
### Ticket
closes #5706 

### What's changed
Add support for NGram speculative decoding

### Checklist
- [X] New/Existing tests provide coverage for changes

### CODE REVIEW

File: integrations/vllm_plugin/vllm_tt/model_runner.py [hunk 1]
```diff
@@ -2105,12 +2410,61 @@ def sample_tokens(
...
+        logits: torch.Tensor,
+        sampling_metadata: XLASupportedSamplingMetadata,
+    ) -> torch.Tensor:
+        """Return one sampling-logit row per request.
+
+        In spec decode, postprocess computes logits for packed draft+bonus
+        positions with shape [batch * (num_spec_tokens + 1), vocab]. Sampling
+        always runs on the bonus row (last packed column) per request.
+        """
+        return logits
```
Comment (alinakhanTT) [inline integrations/vllm_plugin/vllm_tt/model_runner.py]: _extract_sampling_logits returns logits on the very first line, so the whole body below it never runs. Leftover from debugging maybe?

File: integrations/vllm_plugin/vllm_tt/model_runner.py [hunk 2]
```diff
@@ -3182,6 +3650,7 @@ def capture_model(self) -> None:
         """
         Precompile all the subgraphs with possible input shapes.
         """
+        # return
```
Comment (kmabeeTT) [inline integrations/vllm_plugin/vllm_tt/model_runner.py]: commented out return should be dropped

File: tests/integrations/vllm_plugin/sampling/test_speculative_decode.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    return logits
+
+
+def _make_rejection_sampler() -> RejectionSampler:
+    return RejectionSampler(torch.nn.Identity())
+
+
+@pytest.mark.push
+@pytest.mark.cpu
+def test_speculative_greedy_fallback_accepts_all_matching_drafts():
```
Comment (kmabeeTT) [inline tests/integrations/vllm_plugin/sampling/test_speculative_decode.py:28]: Couple test ideas

- These all run with a single request, so the `start = cu[req_idx - 1]` / `flat_idx` offset math that packs multiple requests into one flat tensor never gets exercised which seems like an easy part to get subtly wrong while still passing current tests.   Consider adding batch>1 case with 2 different draft lengths?


- How about `draft_len == 0` case too whichw ould hit `output[req, 0] = bonus` branch in `_rejection_sample_fallback` (currently uncovered).

File: integrations/vllm_plugin/vllm_tt/sampler.py [hunk 4]
```diff
@@ -197,6 +203,11 @@ def sample(
                 sampling_metadata.repetition_penalties,
             )
 
+        if sampling_metadata.all_greedy:
+            return self.greedy_sample(logits)
+
+        assert sampling_metadata.temperature is not None
+
```
Comment (kmabeeTT) [inline integrations/vllm_plugin/vllm_tt/sampler.py:209]: Wasn't sure about why this is here, but I guess it's because of the new call site for the bonus token sampling and an all-greedy request has `temperature=None` which would have fired the assert unless it got moved lower here?

File: tests/integrations/vllm_plugin/sampling/test_speculative_decode.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import numpy as np
+import pytest
+import torch
+from vllm_tt.metadata import XLASupportedSamplingMetadata
+
+from integrations.vllm_plugin.vllm_tt.model_runner import TTModelRunner
+from integrations.vllm_plugin.vllm_tt.rejection_sampler import (
+    _PLACEHOLDER_TOKEN_ID,
+    RejectionSampler,
+)
```
Comment (kmabeeTT) [inline tests/integrations/vllm_plugin/sampling/test_speculative_decode.py]: These imports are causing errors on CI during collection time.... Should we use the `vllm_tt.` root like the other tests here instead of `integrations.vllm_plugin.vllm_tt.` ?

File: integrations/vllm_plugin/vllm_tt/model_runner.py [hunk 6]
```diff
@@ -826,13 +842,25 @@ def _verify_num_xla_graphs(self, case_str):
 
         curr_cached_graph = xr.get_num_cached_compilation_graph()
         new_compiled_graphs = curr_cached_graph - self.num_xla_graphs
+        self.num_xla_graphs += new_compiled_graphs
+
+        # Spec decode runs rejection sampling outside compiled TT graphs, so one
+        # runtime graph may appear during inference.
+        if (
+            self.speculative_config is not None
+            and new_compiled_graphs == 1
```
Comment (kmabeeTT) [inline integrations/vllm_plugin/vllm_tt/model_runner.py]: Hmm eager rejection path does run a few device ops (the `bonus_logits` gather + `self.sampler` call), and their shapes vary with draft length, so I'm not sure the new-graph count is always exactly 1... would >=1 be safer?

File: integrations/vllm_plugin/vllm_tt/model_runner.py [hunk 7]
```diff
@@ -1825,6 +1970,129 @@ def _restore_model_hidden_states(
...
+        """
+        self._draft_token_req_ids = []
+        self._draft_token_ids = []
+
+        if not self.drafter or not self.num_spec_tokens:
+            return
+
+        # Convert sampled_token_ids to list of lists for ngram proposer
+        sampled_token_ids_list = sampled_token_ids.cpu().tolist()
+        if not isinstance(sampled_token_ids_list[0], list):
```
Comment (kmabeeTT) [inline integrations/vllm_plugin/vllm_tt/model_runner.py:2022]: Not sure if valid, but would this ever hit `IndexError` if sampled_token_ids_list is empty? Consider adding to early return unless it's never going to happen.

**Review by kmabeeTT — COMMENTED** [review]
Nice - overall looks great! A few comments and suggestions added

File: integrations/vllm_plugin/vllm_tt/model_runner.py (hunk 1 above)
Comment (mmanzoorTT) [reply integrations/vllm_plugin/vllm_tt/model_runner.py]: Yes, you are right. Claude has added `_extract_sampling_logits` but we don't need this in my opinion. I am running tests with it and will remove it if we don't need it.

File: integrations/vllm_plugin/vllm_tt/sampler.py (hunk 4 above)
Comment (mmanzoorTT) [reply integrations/vllm_plugin/vllm_tt/sampler.py:209]: Yes, this is right.

File: integrations/vllm_plugin/vllm_tt/model_runner.py (hunk 1 above)
Comment (mmanzoorTT) [reply integrations/vllm_plugin/vllm_tt/model_runner.py]: remvoed

File: integrations/vllm_plugin/vllm_tt/model_runner.py (hunk 2 above)
Comment (mmanzoorTT) [reply integrations/vllm_plugin/vllm_tt/model_runner.py]: Removed. thanks for pointing it out.

File: tests/integrations/vllm_plugin/sampling/test_speculative_decode.py (hunk 5 above)
Comment (mmanzoorTT) [reply tests/integrations/vllm_plugin/sampling/test_speculative_decode.py]: Fixed. thanks.

**Review by kmabeeTT — APPROVED** [review]
Thx for changes!

**Review by alinakhanTT — APPROVED** [review]

File: tests/integrations/vllm_plugin/sampling/test_speculative_decode.py (hunk 3 above)
Comment (mmanzoorTT) [reply tests/integrations/vllm_plugin/sampling/test_speculative_decode.py:28]: Thanks for the suggestion. tests added.
